### PR TITLE
fetch bit_depth_* parameters before returning from put

### LINKED ISSUE #59: Allow defining parameters that should be fetched before returning success for a put rather than setting stale parameters
We would like to try fetching `bit_depth_{image,readout}` when it is in the put response before returning.

A hook could be added in [here](https://github.com/DiamondLightSource/fastcs-eiger/blob/main/src/fastcs_eiger/eiger_controller.py#L87) to pop the configured parameters from the list fetch them simmediately (with `controller.<param>.update`). Note this will not be type safe until https://github.com/DiamondLightSource/FastCS/issues/124 is complete.

## Acceptance Criteria
- Specific criteria that will be used to judge if the issue is fixed

### PR DESCRIPTION
**Pull request by jsouter**
Closes #59

### CODE REVIEW

File: tests/system/test_introspection.py [hunk 1]
```diff
@@ -136,3 +137,30 @@ async def test_controller_groups_and_parameters(sim_eiger_controller: EigerContr
...
+    await count_time_attr.updater.put(detector_controller, count_time_attr, 2)
+    assert (
+        detector_controller._parameter_updates
+        and "bit_depth_image" not in detector_controller._parameter_updates
+        and "bit_depth_readout" not in detector_controller._parameter_updates
+    )
+    count_time_attr.updater.config_update.assert_not_called()
+    bit_depth_image_attr.updater.config_update.assert_called()
+    await detector_controller.update()
+    count_time_attr.updater.config_update.assert_called()
```
Comment (GDYendell) [inline tests/system/test_introspection.py:166]: Nit: It is a bit hard to read in one big block. I generally break things up into the initial setup and blocks of function calls with the relevant asserts.

```suggestion
    controller = sim_eiger_controller
    await controller.initialise()
    detector_controller = controller.get_sub_controllers()["Detector"]
    count_time_attr = detector_controller.attributes.get("count_time")
    bit_depth_image_attr = detector_controller.attributes.get("bit_depth_image")
    bit_depth_image_attr.updater.config_update = AsyncMock()
    count_time_attr.updater.config_update = AsyncMock()
    
    assert not detector_controller._parameter_updates
    await count_time_attr.updater.put(detector_controller, count_time_attr, 2)
    assert (
        detector_controller._parameter_updates
        and "bit_depth_image" not in detector_controller._parameter_updates
        and "bit_depth_readout" not in detector_controller._parameter_updates
    )
    count_time_attr.updater.config_update.assert_not_called()
    bit_depth_image_attr.updater.config_update.assert_called()
    
    await detector_controller.update()
    count_time_attr.updater.config_update.assert_called()
```

**Review by GDYendell — APPROVED** [review]
Looks good. I guess this depends on the eiger-stream2 branch of the tickit sim that has the put responses?

File: src/fastcs_eiger/eiger_controller.py [hunk 2]
```diff
@@ -83,7 +85,12 @@ async def put(
...
                 f"Fetching parameters after setting {self.uri}: {parameters_to_update}"
             )
-
+        for param in FETCH_BEFORE_RETURNING:
+            if param in parameters_to_update:
+                print(f"Fetching parameter {param} before returning from put")
+                parameters_to_update.remove(param)
+                attr = controller.attributes.get(_key_to_attribute_name(param))
+                await attr.updater.config_update(controller, attr)
         await controller.queue_update(parameters_to_update)
```
Comment (GDYendell) [inline src/fastcs_eiger/eiger_controller.py:94]: Nit: I would visually separate these steps

```suggestion
                await attr.updater.config_update(controller, attr)
 
        await controller.queue_update(parameters_to_update)
```

File: src/fastcs_eiger/eiger_controller.py [hunk 3]
```diff
@@ -83,6 +85,14 @@ async def put(
...
                 f"Fetching parameters after setting {self.uri}: {parameters_to_update}"
             )
+        for param in FETCH_BEFORE_RETURNING:
+            if param in parameters_to_update:
+                print(f"Fetching parameter {param} before returning from put")
+                parameters_to_update.remove(param)
+                attr_to_update = controller.attributes.get(
+                    _key_to_attribute_name(param)
+                )
+                await attr_to_update.updater.config_update(controller, attr)  # type: ignore
```
Comment (GDYendell) [inline src/fastcs_eiger/eiger_controller.py:95]: What is wrong with the types here? Could we do an assert instead of an ignore at least?

Comment (jsouter) [reply src/fastcs_eiger/eiger_controller.py]: Can do this with 2 asserts, but also noticed I was passing the wrong attr to the config_update! should be fixed now

File: tests/system/test_introspection.py [hunk 4]
```diff
@@ -136,3 +137,37 @@ async def test_controller_groups_and_parameters(sim_eiger_controller: EigerContr
...
+        and "bit_depth_image" not in to_update
+        and "bit_depth_readout" not in to_update
+    )
+    count_time_attr.updater.config_update.assert_not_called()  # type: ignore
+    bit_depth_image_attr.updater.config_update.assert_called()  # type: ignore
+
+    await detector_controller.update()  # type: ignore
+    count_time_attr.updater.config_update.assert_called()  # type: ignore
+
+    await controller.connection.close()
```
Comment (GDYendell) [inline tests/system/test_introspection.py:173]: If you use [mocker.spy](https://pytest-mock.readthedocs.io/en/latest/usage.html#spy) instead of setting to mocks, it might resolve the typing matches.

File: src/fastcs_eiger/eiger_controller.py [hunk 5]
```diff
@@ -83,6 +85,16 @@ async def put(
...
+        for param in FETCH_BEFORE_RETURNING:
+            if param in parameters_to_update:
+                parameters_to_update.remove(param)
+                attr_to_update = controller.attributes.get(
+                    _key_to_attribute_name(param)
+                )
+                assert isinstance(attr_to_update, AttrR)
+                assert isinstance(attr_to_update.updater, EigerConfigHandler)
+                print(f"Fetching parameter {param} before returning from put")
+                await attr_to_update.updater.config_update(controller, attr_to_update)
```
Comment (GDYendell) [inline src/fastcs_eiger/eiger_controller.py:97]: This would be neater with a match statement.

This method is getting a bit big with the different edge cases. Do you think it would be nicer to have a helper method (that takes the `parmeters_to_update` and handles everything after "Manually fetching parameters"?

File: tests/system/test_introspection.py (hunk 4 above)
Comment (jsouter) [reply tests/system/test_introspection.py:173]: Typing errors here are more to do with not knowing the types of the attr and controller, but I've made this change too. Very handy feature!

File: src/fastcs_eiger/eiger_controller.py (hunk 5 above)
Comment (jsouter) [reply src/fastcs_eiger/eiger_controller.py:97]: I'll have a go, maybe I can also move the logic to gather a set of updates from a list of parameter names to a method in the controller so I can avoid reusing the logic in EigerSubsystemController.update and in the EigerHandler.put as well

Comment (jsouter) [reply src/fastcs_eiger/eiger_controller.py:97]: But actually I'm changing how update works in https://github.com/DiamondLightSource/fastcs-eiger/pull/60 anyway so maybe that sort of thing would be better handled there...

Comment (jsouter) [reply src/fastcs_eiger/eiger_controller.py]: Trying to get this working with a match statement but trying to figure out if pyright can handle type narrowing from the member attributes within a pattern match. investigating

Comment (GDYendell) [reply src/fastcs_eiger/eiger_controller.py]: Yeah OK, which one do you think it makes sense to merge first?

**Review by GDYendell — APPROVED** [review]
Happy to merge when you are ready @jsouter
